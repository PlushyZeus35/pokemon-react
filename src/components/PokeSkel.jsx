
export const PokeSkel = () => {
  return (
    <div className="card mb-3">
        <div className="row g-0">
            <div className="col-md-3 d-flex justify-content-center align-items-center">
                <div id="bigSpinner" className="spinner-border" role="status">
                    <span className="visually-hidden">Loading...</span>
                </div>
            </div>
            <div className="col-md-9">
                <div className="card-body">
                    <p className="card-text placeholder-glow">
                        <span className="placeholder col-7"></span>
                        <span className="placeholder col-4"></span>
                        <span className="placeholder col-4"></span>
                        <span className="placeholder col-6"></span>
                        <span className="placeholder col-8"></span>
                    </p>
                </div>
            </div>
        </div>
    </div>
  )
}
