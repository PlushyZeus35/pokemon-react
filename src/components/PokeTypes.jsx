import PropTypes from 'prop-types';

const getIconUrl = type => `/icons/${type}.png`
const generateKey = (pre) => {
    return `${ pre }_${ new Date().getTime() }`;
}
export const PokeTypes = ({types}) => {
    return (
        <div className="d-flex justify-content-center gap-2 my-2">
            {types.map(type => <img key={generateKey(type.type.name)} className="typeImg" src={getIconUrl(type.type.name)} alt={type.type.name}/>)}
        </div>
    )
}

PokeTypes.propTypes = {
    types: PropTypes.arrayOf(PropTypes.object)
}

PokeTypes.defaultProps = {
    types: []
}