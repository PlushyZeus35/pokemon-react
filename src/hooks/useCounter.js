import { useState } from "react"

export const useCounter = (initVal) => {
    const [counter, setCounter] = useState(initVal);

    const increment = () => {
        setCounter(counter+5);
    }

    const decrement = () => {
        if(counter==0) return;
        setCounter(counter-5);
    }

    return {
        counter,
        increment,
        decrement
    }
}
