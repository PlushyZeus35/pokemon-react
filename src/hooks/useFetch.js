import { useEffect, useState } from "react"

export const useFetch = (url) => {
	const [state, setState] = useState({
		data: null,
		isLoading: true,
		isError: null
	})

	const getPokemons = async () => {
		setState({data: null, isLoading: true, isError: null})
		const resp = await fetch(url);
		const respData = await resp.json();
		setState({data: respData, isLoading: false, isError: null})
	}

	useEffect(() => {
		getPokemons();
	}, [url])
	

	return {
		data: state.data,
		isLoading: state.isLoading,
		isError: state.isError
	}
}
