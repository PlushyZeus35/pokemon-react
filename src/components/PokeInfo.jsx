import { PokeTypes } from "./PokeTypes"
import PropTypes from 'prop-types';

export const PokeInfo = ({name, img, types}) => {
  return (
    <div className="card mb-3">
        <div className="row g-0">
            <div className="col-md-3 d-flex justify-content-center">
                <img src={img} className="img-fluid rounded-start" alt={name}/>
            </div>
            <div className="col-md-9">
                <div className="card-body d-flex flex-column justify-content-center">
                    <h5 className="card-title text-center text-capitalize">{name}</h5>
                    <PokeTypes types={types}/>
                    <button className="btn btn-primary">Ver m√°s</button>
                </div>
            </div>
        </div>
        </div>
  )
}

PokeInfo.propTypes = {
    name: PropTypes.string.isRequired,
    img: PropTypes.string.isRequired,
    types: PropTypes.arrayOf(PropTypes.object)
}

PokeInfo.defaultProps = {
    types: []
}