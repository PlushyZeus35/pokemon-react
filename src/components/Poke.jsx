import { useFetch } from "../hooks/useFetch"
import { PokeInfo } from "./PokeInfo";
import { PokeSkel } from "./PokeSkel";
import PropTypes from 'prop-types';

export const Poke = ({name, url}) => {

    const {data, isLoading, isError} = useFetch(url);
    const {sprites, types} = isLoading ? {} : data;

    return (
        <>
            {
                isLoading && (<PokeSkel />)
            }
            {
                isLoading || (<PokeInfo name={name} img={sprites?.front_default} types={types} />)
            }
        </>
    )
}

Poke.propTypes = {
    name: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}