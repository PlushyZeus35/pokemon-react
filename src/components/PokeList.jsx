import { useContext } from "react"
import { useCounter } from "../hooks/useCounter"
import { useFetch } from "../hooks/useFetch";
import { LoadingList } from "./LoadingList";
import { Poke } from "./Poke";

export const PokeList = () => {

    const {counter, increment, decrement} = useCounter(0);
    const {isLoading, data, isError} = useFetch(`https://pokeapi.co/api/v2/pokemon?offset=${counter}&limit=5`);
    const {results} = isLoading ? [] : data;
    return (
        <div>
            <div id="pokeList">
                
                    {
                        isLoading || 
                        results.map(poke => <Poke key={poke.url} name={poke.name} url={poke.url} />)
                    }
                    {
                        isLoading && 
                        <LoadingList key="loading" />
                    }
                
            </div>
            <div id="buttonsContainer" className="d-flex justify-content-center">
                <div className="btn-group " role="group" aria-label="Button group">
                    <button data-testid="dec-page" type="button" onClick={decrement} className="btn btn-lg btn-primary">
                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" className="bi bi-caret-left-fill" viewBox="0 0 16 16">
                            <path d="m3.86 8.753 5.482 4.796c.646.566 1.658.106 1.658-.753V3.204a1 1 0 0 0-1.659-.753l-5.48 4.796a1 1 0 0 0 0 1.506z"/>
                        </svg>
                    </button>
                    <button data-testid="page-counter" type="button" onClick={decrement} className="btn btn-lg btn-primary">
                        {(counter/5)+1}
                    </button>
                    <button data-testid="inc-page" type="button" onClick={increment} className="btn btn-lg btn-primary">
                        <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" className="bi bi-caret-right-fill" viewBox="0 0 16 16">
                            <path d="m12.14 8.753-5.482 4.796c-.646.566-1.658.106-1.658-.753V3.204a1 1 0 0 1 1.659-.753l5.48 4.796a1 1 0 0 1 0 1.506z"/>
                        </svg>
                    </button>
                </div>
            </div>
        </div>
    )
}
