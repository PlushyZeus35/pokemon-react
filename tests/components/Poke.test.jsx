import { render, screen } from "@testing-library/react"
import { Poke } from "../../src/components/Poke"
import { useFetch } from "../../src/hooks/useFetch"

// Set mock function to useFetch hook
jest.mock('../../src/hooks/useFetch');

describe('Tests of <Poke /> component', () => {
    const pokeName = 'charizard';
    const pokeUrl = 'https://testurl.com/';
    const pokeInfo = {
        name: pokeName,
        types: [{type: {name: "water"}}],
        sprites: {
            front_default: 'https://testurlimg.com/'
        }
    }
    test('Should render loading initially', ()=>{
        useFetch.mockReturnValue({
            data: [],
            isLoading: true,
            isError: false
        })
        render(<Poke name={pokeName} url={pokeUrl}/>)
        expect(screen.getByText('Loading...')).toBeTruthy();
    })

    test('Should render pokemon after receiving data from hook', ()=>{
        useFetch.mockReturnValue({
            data: pokeInfo,
            isLoading: false,
            isError: false
        })
        const {container} = render(<Poke name={pokeName} url={pokeUrl}/>)
        expect(screen.getAllByRole('img').length).toBe(2);
        expect(screen.getByAltText(pokeName).src).toBe(pokeInfo.sprites.front_default);
        expect(container).toMatchSnapshot();
    })
})