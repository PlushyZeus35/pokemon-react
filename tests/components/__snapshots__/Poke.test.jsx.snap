// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of <Poke /> component Should render pokemon after receiving data from hook 1`] = `
<div>
  <div
    class="card mb-3"
  >
    <div
      class="row g-0"
    >
      <div
        class="col-md-3 d-flex justify-content-center"
      >
        <img
          alt="charizard"
          class="img-fluid rounded-start"
          src="https://testurlimg.com/"
        />
      </div>
      <div
        class="col-md-9"
      >
        <div
          class="card-body d-flex flex-column justify-content-center"
        >
          <h5
            class="card-title text-center text-capitalize"
          >
            charizard
          </h5>
          <div
            class="d-flex justify-content-center gap-2 my-2"
          >
            <img
              alt="water"
              class="typeImg"
              src="/icons/water.png"
            />
          </div>
          <button
            class="btn btn-primary"
          >
            Ver m√°s
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
